src/python/pants/engine/process.py

```diff
@@ -50,6 +51,57 @@ class ProcessCacheScope(Enum):
     PER_SESSION = "per_session"
 
 
+@dataclass(frozen=True)
+class ProcessConcurrency:
+    kind: Literal["exactly", "range", "exclusive"]
+    min: int | None = None
+    max: int | None = None
+
+    def __post_init__(self):
+        if self.min is not None and self.min < 1:
+            raise ValueError(f"min concurrency must be >= 1, got {self.min}")
+        if self.max is not None and self.max < 1:
+            raise ValueError(f"max concurrency must be >= 1, got {self.max}")
+        if self.min is not None and self.max is not None and self.min > self.max:
+            raise ValueError(
+                f"min concurrency must be <= max concurrency, got {self.min} and {self.max}"
+            )
+        if self.kind == "exactly" and self.min != self.max:
+            raise ValueError(
+                f"exactly concurrency must have min and max equal, got {self.min} and {self.max}"
+            )
+
+    @staticmethod
+    def range(max: int, min: int = 1):
+        """The amount of parallelism that this process is capable of given its inputs. This value
```

Comment:
copied from `concurrency_available` in rust. I wanted it in py so its more discoverable